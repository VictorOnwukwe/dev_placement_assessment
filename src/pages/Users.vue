<template>
  <div style="position:relative;z-index:2">
    <div
      v-if="users.length == 0"
      style="display:flex;justify-content:center;align-items:center;height:100%"
    >
      <div class="loader"></div>
    </div>
    <div v-else class="display">
      <preview
        v-for="user in currentUsers"
        :key="user.email"
        :user="user"
      ></preview>
    </div>
  </div>
</template>
<script>
import UserPreview from "../components/UserPreview";
export default {
  components: {
    preview: UserPreview,
  },
  props: {
    gender: {
      type: String,
      default: "all",
    },
  },
  data() {
    return {
      currentUsers: [],
      displayUsers: [],
    };
  },
  methods: {
    updateCurrentUsers() {
      let start = (this.page - 1) * 3;
      let end = start + 3;
      this.currentUsers = this.displayUsers.slice(start, end);
    },
    filterByGender() {
      let val = this.gender;
      if (val == "male") {
        this.displayUsers = this.users.filter((user) => user.gender == "male");
      } else if (val == "female") {
        this.displayUsers = this.users.filter(
          (user) => user.gender == "female"
        );
      } else {
        this.displayUsers = this.users;
      }
    },
  },
  computed: {
    page() {
      return this.$store.getters["users/currentPage"];
    },
    users() {
      return this.$store.getters["users/allUsers"];
    },
  },
  watch: {
    page() {
      this.updateCurrentUsers();
    },
    gender() {
      this.$store.dispatch("users/setCurrentPage", 1);
      this.filterByGender();
      this.updateCurrentUsers();
    },
    users() {
      this.filterByGender();
      this.updateCurrentUsers();
    },
  },
  mounted() {
    this.filterByGender();
    this.updateCurrentUsers();
  },
};
</script>
<style scoped>
.display *:nth-child(n + 2) {
  margin-top: 16px;
}
.absolute {
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  z-index: ;
}
.loader {
  width: 40px;
  height: 40px;
  border-radius: 50%;
  border-top: 4px solid #f7f7ff;
  border-right: 4px solid #f7f7ff;
  border-bottom: 4px solid #e2e3ec;
  border-left: 4px solid #e2e3ec;
  animation: rotate 0.5s linear infinite;
}
@keyframes rotate {
  0% {
    transform: rotate(0);
  }
  100% {
    transform: rotate(360deg);
  }
}
</style>
