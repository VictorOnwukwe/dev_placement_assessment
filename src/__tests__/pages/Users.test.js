import { mount, config, createLocalVue } from "@vue/test-utils";
import Users from "../../pages/Users";
import Vuex from "vuex";

config.showDeprecationWarnings = false;

describe("Users Display Page", () => {
  let allUsers = Array(20)
    .fill()
    .map((v, i) => {
      return {
        name: { first: "i", last: "last" },
        picture: { medium: "medium" },
        location: {
          street: { number: "1", name: "name" },
          city: "city",
          state: "state",
          country: i % 4 == 0 ? "Nigeria" : "other",
        },
        email: "email" + i,
        phone: "112233",
        gender: i % 3 == 0 ? "male" : "female",
      };
    });
  let defaultData;
  const localVue = createLocalVue();
  localVue.use(Vuex);
  let store;

  beforeEach(() => {
    store = new Vuex.Store({
      getters: {
        allUsers: () => allUsers,
      },
    });

    defaultData = {
      store,
      mocks: {
        $route: { params: { email: "email" } },
      },
      localVue,
      stubs: ["router-link"],
      methods: {
        updateContent: jest.fn()
      }
    };
  });

  describe("Sort By Gender", () => {
    let wrapper, store;
    let maleUsers = allUsers.filter((user) => user.gender == "male");
    let femaleUsers = allUsers.filter((user) => user.gender == "female");
    beforeEach(() => {
      store = new Vuex.Store({
        getters: {
          allUsers: () => allUsers,
        },
      });
      defaultData = { ...defaultData, store };
    });
    afterEach(() => {
      wrapper.destroy();
    });
    it("should return populate users with gender neutral array when gender is 'all'", () => {
      wrapper = mount(Users, {
        ...defaultData,
        propsData: {
          gender: "all",
          country: "all",
          search: "",
        },
      });

      expect(wrapper.vm.$data.displayUsers).toEqual(allUsers);
    });

    it("should populate display users with male gender array when gender is 'male'", () => {
      wrapper = mount(Users, {
        ...defaultData,
        propsData: {
          gender: "male",
          country: "all",
          search: "",
        },
      });

      expect(wrapper.vm.$data.displayUsers).toEqual(maleUsers);
    });

    it("should populate display users with female gender array when gender is 'female'", () => {
      wrapper = mount(Users, {
        ...defaultData,
        propsData: {
          gender: "female",
          country: "all",
          search: "",
        },
      });

      expect(wrapper.vm.$data.displayUsers).toEqual(femaleUsers);
    });
  });

  describe("Sort By Country", () => {
    let wrapper, store;
    let nigerianUsers = allUsers.filter((u) => u.location.country == "Nigeria");
    afterEach(() => {
      wrapper.destroy();
    });
    it("should populate displayUsers with users from a specific country", () => {
      wrapper = mount(Users, {
        ...defaultData,
        propsData: {
          gender: "all",
          country: "Nigeria",
          search: "",
        },
      });

      expect(wrapper.vm.$data.displayUsers).toEqual(nigerianUsers);
    });
  });

  describe("Search", () => {
    let matches = (user, search, showCountry) => {
      let userString = `${user.location.street.number} ${
        user.location.street.name
      } ${user.location.city}, ${user.location.state}${
        showCountry ? " " + user.location.country : ""
      } ${user.name.first} ${user.name.last} ${user.email}`.toLocaleLowerCase();

      return userString.includes(search.trim().toLowerCase());
    };
    let searchResults;

    it("should populate displayUsers with users that match keyword when country is shown", async () => {
      searchResults = allUsers.filter((user) => matches(user, "i", true));
      const wrapper = mount(Users, {
        ...defaultData,
        propsData: {
          gender: "all",
          country: "all",
          search: "i",
        },
        store: new Vuex.Store({
          getters: {
            showCountry: () => true,
            allUsers: () => allUsers,
          },
        }),
      });

      expect(wrapper.vm.$data.displayUsers).toEqual(searchResults);
    });
    it("should populate displayUsers with users that match keyword when country is hidden", async () => {
      searchResults = allUsers.filter((user) => matches(user, "i", false));
      const wrapper = mount(Users, {
        ...defaultData,
        propsData: {
          gender: "all",
          country: "all",
          search: "i",
        },
        store: new Vuex.Store({
          getters: {
            showCountry: () => false,
            allUsers: () => allUsers,
          },
        }),
      });

      expect(wrapper.vm.$data.displayUsers).toEqual(searchResults);
    });
  });

  it("should return data according to page number", () => {
    store = new Vuex.Store({
      getters: {
        allUsers: () => allUsers,
        currentPage: () => 3,
      },
    });
    const wrapper = mount(Users, {
      ...defaultData,
      store,
      propsData: {
        gender: "all",
        country: "all",
        search: "",
      },
      methods: {}
    });

    expect(wrapper.vm.$data.currentUsers).toEqual(allUsers.slice(6, 9));
  });

  it("should render correctly when passed complete data", () => {
    const wrapper = mount(Users, {...defaultData});
    expect(wrapper.html()).toMatchSnapshot();
  });
});
