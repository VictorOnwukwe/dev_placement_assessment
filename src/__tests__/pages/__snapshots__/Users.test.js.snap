// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Users Display Page should render correctly when passed complete data 1`] = `
<div style="position: relative; z-index: 2;">
  <div style="display: flex; justify-content: center; align-items: center; height: 100%;">
    <div class="loader"></div>
  </div>
</div>
`;
